scala/scala-impl/src/org/jetbrains/plugins/scala/lang/psi/impl/expr/ScForImpl.scala

```diff
@@ -183,8 +183,15 @@ class ScForImpl(node: ASTNode) extends ScExpressionImplBase(node) with ScFor wit
       normalizeUnderscores(expr).getText
     }
 
-    def needsPatternMatchFilter(pattern: ScPattern): Boolean =
-      !pattern.isIrrefutableFor(if (forDisplay) pattern.expectedType else None)
+    def needsPatternMatchFilter(pattern: ScPattern): Boolean = {
+      val hasSc3Case = pattern.isScala3OrSource3Enabled &&
+        pattern.prevSiblingNotWhitespace.exists(_.getText == ScalaModifier.CASE)
```

Comment:
As IntelliJ suggests, please replace `_.getText == ScalaModifier.CASE` with `_.textMatches(ScalaModifier.CASE)`

Reply:
Ah, right the inspection should work in internal mode.
https://github.com/JetBrains/intellij-scala/blob/7a050b30da86b846ca88fa92f51f875b7b805c1d/README.md#prerequisites

> (optional but recommended)
Enable [internal mode](https://plugins.jetbrains.com/docs/intellij/enabling-internal.html) in IDEA to get access to helpful internal actions and debug information